src/output/plugins/visualization/SoundAnalysis.hxx

```diff
@@ -0,0 +1,317 @@
+#include <cstddef>
+#include <memory>
+
+#include <math.h>
+
+struct ConfigBlock;
+
+namespace Visualization {
+
+/**
+ * \brief Convenience class for expressing sound analysis parameters exclusive
+ * of the audio format
+ *
+ *
+ * There are any number of parameters governing our analysis of PCM data. Other
+ * than the `AudioFormat`, they are read from configuration at startup and
+ * constant. Rather than force callers to write methods taking many parameters,
+ * this class colects them all in one place, and enforces some constraints on
+ * their values.
+ *
+ *
+ */
+
+class SoundAnalysisParameters
+{
+	/* The number of samples used for each analysis; this must be greater
+	 * than zero and needn't be large (say, less than 1024). Configuration
+	 * value "num_samples" */
+	size_t num_samples;
+	/* Data lower than this frequency (in the frequency domain) shall be
+	 * discarded; must be greater than or equal to zero, and less than
+	 * hi_cutoff. A typical value would be 200 (the lower range of human
+	 * perception). Units: Hz. Configuration value "lo_cutoff" */
+	float lo_cutoff;
+	/* Data greater than this frequency (in the frequency domain) shall be
+	 * discarded; must be greater than or equal to zero, and greater than
+	 * lo_cutoff. A typical value would be 10000-12000 (empirically, there's
+	 * not a lot of activity above 10000 in song data). Units
+	 * Hz. Configuration value "hi_cutoff" */
+	float hi_cutoff;
+
+	static constexpr size_t DEFAULT_NUM_SAMPLES = 513;
+	static constexpr size_t DEFAULT_LO_CUTOFF = 200;
+	static constexpr size_t DEFAULT_HI_CUTOFF = 10000;
+
+public:
+	SoundAnalysisParameters() noexcept;
+	explicit SoundAnalysisParameters(const ConfigBlock &config_block);
+	SoundAnalysisParameters(size_t num_samples, float lo_cutoff,
+				float hi_cutoff);
+
+	size_t
+	GetNumSamples() const noexcept {
+		return num_samples;
+	}
+	float
+	GetLoCutoff() const noexcept {
+		return lo_cutoff;
+	}
+	float
+	GetHiCutoff() const noexcept {
+		return hi_cutoff;
+	}
+};
+
+/**
+ * \class SoundAnalysis
+ *
+ * \brief Analayze PCM data in a manner convienient for visualization authors
+ *
+ *
+ * This class houses our logic for going from raw PCM data to the power
+ * spectrum, bass/mids/trebs &c. Instances are constructed with configuration
+ * information on the analysis details, and repeated analysis for different
+ * datasets is performed via update(). Since instances allocate input & output
+ * buffers for the discrete Fourier transform, they are not meant to be copied
+ * around.
+ *
+ *
+ */
+
+class SoundAnalysis {
+
+	/// # of samples to be used in each analysis
+	size_t num_samples;
+	/* # of Fourier coefficients computed by FFTW (should be
+	 * (num_samples / 2) + 1) */
+	size_t out_samples;
+	std::shared_ptr<SoundInfoCache> pcache;
+	AudioFormat audio_format;
+	/* # of audio channels (e.g. 1 is mono, 2 is stereo--
+	 * # cf. SampleFormat.hxx); should be audio_format.num_channels */
+	uint8_t num_channels;
```

Comment:
Are you aware that this `uint8_t` field in fact allocates 8 bytes in this struct?